From d7ba71efa64b2ff790d7c7e50762740dda38e773 Mon Sep 17 00:00:00 2001
From: Igal Zeltser <izeltser@habana.ai>
Date: Wed, 02 Nov 2022 10:07:48 +0200
Subject: [PATCH] [SW-0] gaudi3: enable lbw early write response

Early write response is boosting CPU performance by generating "fake"
write transaction acknowledge before the actual terminator receives
the transaction. This allow to better utilize the data pipeline of the
NOC.

Change-Id: Id09139595844c77249f4bf34d3ae6f1eba1524b2
Signed-off-by: Igal Zeltser <izeltser@habana.ai>
---

diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index 81a00ab..ed3844d 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -21,6 +21,8 @@
 #include <d0_parc_int_aggr_i2c_slv_comb_regs.h>
 #include <d0_parc_local_cfg_regs.h>
 #include <d0_pif_arc_regs.h>
+#include <d0_parc_global_conf_regs.h>
+#include <d0_parc_global_conf_masks.h>
 #include <d0_pif_arc_special_regs.h>
 #include <d0_g4_pci_pll_ctrl_regs.h>
 #include <d1_pcie_aux_regs.h>
@@ -134,8 +136,23 @@
 	regmap_writel(0x1, mmD0_PSOC_GLOBAL_CONF_PHY_STABLE);
 }
 
+#define EARLY_WRITE_RESP_MASK	\
+	(1 << D0_PARC_GLOBAL_CONF_ARC_LBU_AXI_SPLIT_CTRL2_AWCACHE_OVRD_EN_S | \
+	 1 << D0_PARC_GLOBAL_CONF_ARC_LBU_AXI_SPLIT_CTRL2_ARCACHE_OVRD_EN_S | \
+	 D0_PARC_GLOBAL_CONF_ARC_LBU_AXI_SPLIT_CTRL2_FORCE_WR_BUF_M)
+
+static void hl_lbw_early_write_resp_enable(void)
+{
+	u32 axi_split_val, val;
+
+	axi_split_val = regmap_readl(mmD0_PARC_GLOBAL_CONF_ARC_LBU_AXI_SPLIT_CTRL2_1);
+	val = axi_split_val | EARLY_WRITE_RESP_MASK;
+	regmap_writel(val, mmD0_PARC_GLOBAL_CONF_ARC_LBU_AXI_SPLIT_CTRL2_1);
+}
+
 void hl_plat_early_init(void)
 {
+	hl_lbw_early_write_resp_enable();
 	init_prstn();
 }
 
